File: day06/others/day06.zig
const std = @import("std");
const print = std.debug.print;

fn is_matcher(input: []const u8, k: usize) bool {
    var freq: [26]u8 = [_]u8{0} ** 26;
    for (input) |c| {
        const idx = c - 'a';
        freq[idx] += 1;
    }
    var count: usize = 0;
    for (freq) |f| {
        if (f > 0) count += 1;
    }
    return count == k;
}

fn lookup(input: []const u8, k: usize) i32 {
    var left: usize = 0;
    var right: usize = 0;
    while (right < input.len) {
        right += 1;
        if (is_matcher(input[left..right], k)) {
            return @intCast(i32, right);
        }
        if (right - left >= k) {
            left += 1;
        }
    }
    return -1;
}

pub fn main() void {
    const input = @embedFile("../input.txt");
    print("Part 1 = {}\n", .{lookup(input, 4)});
    print("Part 2 = {}\n", .{lookup(input, 14)});
}

File: day06/others/day06.c
#include <stdio.h>

int is_marker(char* input, int left, int right, int k) {
	int freq[26] = {0};
	for (int i = left; i < right; i++) {
		freq[input[i] - 'a']++;
	}
	int count = 0;
	for (int i = 0; i < 26; i++) {
		if (freq[i] == 1) count++;
	}
	return count == k;
}

int lookup(char* input, int len, int k) {
	int left = 0;
	int right = 0;
	while (right < len) {
		right++;
		if (is_marker(input, left, right, k)) {
			return right;
		}
		if (right - left >= k) {
			left++;
		}
	}
	return -1;
}

int main() {
	FILE* fp = fopen("../input.txt", "r");
	fseek(fp, 0, SEEK_END);
	long length = ftell(fp);
	fseek(fp, 0, SEEK_SET);
	char buf[length];
	fgets(buf, length, fp);

	printf("Part 1 - Found at %d\n", lookup(buf, length, 4));
	printf("Part 2 - Found at %d", lookup(buf, length, 14));

	fclose(fp);
}
